### PR TITLE
pretty print lazy op per default

### PR DESCRIPTION
**Pull request by DKormann**
pretty print LazyOp per default:
```python
t = Tensor.arange(10)

ast = t.schedule()[0].ast

print(ast)
```
```
LazyOp(MetaOps.SINK, arg=None, src=(
  LazyOp(BufferOps.STORE, arg=MemBuffer(idx=0, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(10, 1), strides=(1, 0), offset=0, mask=None, contiguous=True),))), src=(
    LazyOp(BinaryOps.ADD, arg=None, src=(
      LazyOp(ReduceOps.SUM, arg=(1,), src=(
        LazyOp(BufferOps.CONST, arg=ConstBuffer(val=1, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(11, 19), strides=(0, 0), offset=0, mask=((0, 11), (9, 19)), contiguous=False), View(shape=(10, 10), strides=(1, 20), offset=0, mask=None, contiguous=False)))), src=()),)),
      LazyOp(BufferOps.CONST, arg=ConstBuffer(val=-1, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(10, 1), strides=(0, 0), offset=0, mask=None, contiguous=False),))), src=()),)),)),))
```

### CODE REVIEW

**Comment by chenyuxyz** [comment]
it's definitely golfing too much if you need to remove the spaces between words, i think this works
```
  def __repr__(self:LazyOp, indent=0):
    return f"{' '*indent}LazyOp({self.op}, arg={self.arg}, src=({''.join(f'\n{x.__repr__(indent+2)},' for x in self.src)}))"
```

File: tinygrad/ops.py [hunk 1]
```diff
@@ -62,7 +62,8 @@ def cached_compare(self, x, context):
     ret = context[key] = all(a.cached_compare(b, context) for a,b in zip(self.src, x.src))
     return ret
   def __eq__(self, x): return self.cached_compare(x, context={})
-  def __repr__(self): return f"LazyOp(op={self.op}, src={self.src}, arg={self.arg})"
+  def __repr__(self:LazyOp,indent=0):
+    return' '*indent+f"LazyOp({self.op}, arg={self.arg}, src=("+(('\n'+'\n'.join(x.__repr__(indent+2)+','for x in self.src)) if self.src else '')+"))"
```
Comment (geohot) [inline tinygrad/ops.py:66]: Beware of exponential behavior

**Comment by chenyuxyz** [comment]
we discussed offline, fine to merge this after you make the change i suggested, and add a simple test to print something

we will deal with exponential behavior separately as it is also the case in existing code. ultimately we want to make print_tree printing valid python because print_tree solves reference, and we can use walrus operator

**Comment by DKormann** [comment]
just saw your comments I kinda like the walrus version but can revert

**Comment by DKormann** [comment]
the existing TestLazyOp::test_lazyop_str is good already i feel like?

```python
(L0:=LazyOp(BinaryOps.ADD, arg=None, src=(
  (L1:=LazyOp(BinaryOps.ADD, arg=None, src=(
    (L2:=LazyOp(BinaryOps.ADD, arg=None, src=(
      (L3:=LazyOp(BinaryOps.ADD, arg=None, src=(
        (L4:=LazyOp(BinaryOps.ADD, arg=None, src=(
          (L5:=LazyOp(BufferOps.CONST, arg=ConstBuffer(val=1.0, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(1, 1), strides=(0, 0), offset=0, mask=None, contiguous=True),))), src=())),
          (L5:=LazyOp(BufferOps.CONST, arg=ConstBuffer(val=1.0, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(1, 1), strides=(0, 0), offset=0, mask=None, contiguous=True),))), src=()))))),
        L4))),
      L3))),
    L2))),
  L1)))
```

**Comment by DKormann** [comment]
if you want to print_tree to be executable you can just replace it completely with something like this

**Comment by chenyuxyz** [comment]
only want the referenced ones to have walrus

**Comment by DKormann** [comment]
reverted back to simpler version for now

**Comment by DKormann** [comment]
line diff before delete print_tree:

Changes
Name                            Lines    Diff    Tokens/Line    Diff
----------------------------  -------  ------  -------------  ------
tinygrad/codegen/uopgraph.py      482      +4           21.6    -0.0
tinygrad/helpers.py               282      +9           16.0    +0.1
tinygrad/ops.py                   162      +0           16.5    +0.1
tinygrad/codegen/uops.py          161      -1           16.0    +0.2


total lines changes: +12

**Comment by DKormann** [comment]
pretty print lazyop uop upat per default:
```python
 LazyOp(BinaryOps.ADD, arg=None, src=(
   x0:=LazyOp(BinaryOps.ADD, arg=None, src=(
     x1:=LazyOp(BinaryOps.ADD, arg=None, src=(
       x2:=LazyOp(MetaOps.CONST, arg=None, src=(None)),
       x2,)),
     x1,)),
   x0,))
 UOp(UOps.ALU, None, arg=BinaryOps.ADD, src=(
   x0:=UOp(UOps.CONST, None, arg=None, src=()),
   x0,))
 UPat({UOps.ALU}, BinaryOps.ADD, name='None', dtype=None, src=(
   UPat({UOps.CONST}, None, name='None', dtype=None, src=(None)),
   UPat({UOps.CONST}, None, name='None', dtype=None, src=(None)),))
```

**Comment by geohot** [comment]
Please focus on readability instead of line count.

Do we need to print UPat?

File: tinygrad/codegen/uopgraph.py [hunk 2]
```diff
@@ -38,6 +38,10 @@ def compile(u: UOp, name:Optional[str]=None) -> UPat:
     if u.op is UOps.VAR: return UPat(name=name or u.arg, dtype=u.dtype) if len(u.src) == 0 else UPat.compile(u.src[0], name or u.arg)
     return UPat(u.op, u.arg, (list if u.commutative() else tuple)([UPat.compile(src) for src in u.src]) if u.src != () else None,
                 name, u.dtype, allow_any_len=(isinstance(name, str) and 'allow_any_len' in name))
+  def __repr__(self):
+    def rep(x): return f"UPat({{{{{', '.join(str(x) for x in x.op) if isinstance(x.op, tuple) else x.op}}}}}, {x.arg}, name='{x.name}',\
```
Comment (geohot) [inline tinygrad/codegen/uopgraph.py:42]: unreadable.

**Comment by geohot** [comment]
Overall I like it though.
```
diane@blak:~/tinygrad$ python3
Python 3.11.6 (main, Nov 13 2023, 14:32:17) [Clang 15.0.0 (clang-1500.0.40.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from tinygrad.codegen.uops import UOp; from tinygrad import dtypes
>>> a = UOp.const(dtypes.int, 2)
>>> a
 UOp(UOps.CONST, dtypes.int, arg=2, src=())
>>> a/2
 UOp(UOps.ALU, dtypes.int, arg=BinaryOps.MUL, src=(
   x0:=UOp(UOps.CONST, dtypes.int, arg=2, src=()),
   UOp(UOps.ALU, dtypes.int, arg=UnaryOps.RECIP, src=(
     x0,)),))
>>> a+3
 UOp(UOps.ALU, dtypes.int, arg=BinaryOps.ADD, src=(
   UOp(UOps.CONST, dtypes.int, arg=2, src=()),
   UOp(UOps.CONST, dtypes.int, arg=3, src=()),))
```

**Comment by DKormann** [comment]
i really like printing upat for debugging rules but the class is messy and the print function is a mess

File: tinygrad/codegen/uopgraph.py [hunk 3]
```diff
@@ -47,7 +53,7 @@ def _match(uop:UOp, pat:UPat, store:Dict[str, UOp]) -> List[Dict[str, UOp]]:
   if pat.src is None: return [store]
   res: List[Dict[str, UOp]] = []
   for vp in pat.src:
-    if pat.allowed_len != 0 and len(uop.src) != pat.allowed_len: return []
+    if not pat.allow_any_len and len(uop.src) != len(vp): return []
```
Comment (geohot) [inline tinygrad/codegen/uopgraph.py:56]: Why did this change? It's slower

**Pull request merged**